x/stake/keeper/val_state_change.go

```diff
@@ -130,6 +130,92 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx sdk.Context) (newVals []ty
 	return newVals, updates
 }
 
+// update staked tokens snapshots of all validators
+// elect topN validators according to the accumulated staked tokens over snapshotNum
+func (k Keeper) UpdateAndElectValidators(ctx sdk.Context) (newVals []types.Validator, updates []abci.ValidatorUpdate) {
+	snapshotNum := int(k.MaxStakeSnapshots(ctx))
+	var validators []types.Validator
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, ValidatorsKey)
+	defer iterator.Close()
+	var validator types.Validator
+	for ; iterator.Valid(); iterator.Next() {
+		validator = types.MustUnmarshalValidator(k.cdc, iterator.Value())
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+			continue
+		}
+		validator.StakeSnapshots = append(validator.StakeSnapshots, validator.Tokens)
+		snapshotsLen := len(validator.StakeSnapshots)
+		if snapshotsLen > snapshotNum {
+			validator.StakeSnapshots = validator.StakeSnapshots[snapshotsLen-snapshotNum:]
+		}
+		accmulatedStake := sdk.ZeroDec()
+		for _, s := range validator.StakeSnapshots {
+			accmulatedStake = accmulatedStake.Add(s)
+		}
+		validator.AccumulatedStake = accmulatedStake
+		// set snapshot change
+		k.SetValidator(ctx, validator)
+		// validator which is jailed or has no token will not be election candidate
+		if validator.Jailed || validator.Tokens.IsZero() {
+			continue
+		}
+		validators = append(validators, validator)
+	}
+	sort.SliceStable(validators, func(i, j int) bool {
+		return validators[i].AccumulatedStake.GT(validators[j].AccumulatedStake)
```

Comment:
For the same `AccumulatedStake`, let us sort by `OperatorAddr` further, to 100% ensure that the sort result is fixed.

Reply:
fixed